src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,462 @@
+    replaceable = ""True""
+  />
+</root>";
+        }
+
+        internal bool CompilationNeeded(HashSet<string> files, string outputFileName)
+        {
+            if (!_fileSystem.FileExists(outputFileName))
+                return true;
+
+            DateTime outputDateTime = _fileSystem.LastFileWriteTime(outputFileName);
+
+            foreach (string file in files)
+            {
+                DateTime fileDateTime = _fileSystem.LastFileWriteTime(file);
+                if (fileDateTime > outputDateTime)
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static string GetOutputFileName(string outputPath, string fileName)
+        {
+            return Path.Combine(outputPath, fileName.Replace('\\', '.') + ".dll");
+        }
+
+        protected virtual Task CompileTempPEAsync(HashSet<string> filesToCompile, string outputFileName)
+        {
+            return _sequentialTaskQueue.ExecuteTask(async () =>
+            {
+                bool result = await _compiler.CompileAsync(_languageServiceHost.ActiveProjectContext, outputFileName, filesToCompile, default);
+
+                // if the compilation failed or was cancelled we should clean up any old TempPE outputs lest a designer gets the wrong types, plus its what legacy did
+                if (!result)
+                {
+                    try
+                    {
+                        if (_fileSystem.FileExists(outputFileName))
+                        {
+                            _fileSystem.RemoveFile(outputFileName);
+                        }
+                    }
+                    catch (IOException)
+                    { }
+                }
+            });
+        }
+
+        protected virtual HashSet<string> GetFilesToCompile(string fileName, ImmutableHashSet<string> sharedInputs)
+        {
+            // This is a HashSet because we allow files to be both inputs and shared inputs, and we don't want to compile the same file twice
+            // plus Roslyn needs to call Contains on this quite a lot in order to ensure its only compiling the right files so we want that to be fast.
+            var files = new HashSet<string>(sharedInputs.Count + 1, StringComparers.Paths);
+            files.AddRange(sharedInputs.Select(UnconfiguredProject.MakeRooted));
+            files.Add(UnconfiguredProject.MakeRooted(fileName));
+            return files;
+        }
+
+        /// <summary>
+        /// ApplyAsync is called on the UI thread and its job is to update AppliedValue to be correct based on the changes that have come through data flow after being processed
+        /// </summary>
+        protected override async Task ApplyAsync(DesignTimeInputsDelta value)
+        {
+            // Not using use the ThreadingService property because unit tests
+            _unconfiguredProjectServices.ThreadingService.VerifyOnUIThread();
+
+            DesignTimeInputsItem previousValue = AppliedValue.Value;
```

Comment:
I can't find this type?

Reply:
This, and `DesignTimeInputsDelta` are at the bottom of the file. Seemed to be the convention of other ProjectHostBridge classes.

Reply:
Can we split them out into a separate file (but nested), to match tree convention.